### PR TITLE
KDESKTOP-875-Upload-of-big-folder-very-slow-with-LiteSync-activated

### PR DESCRIPTION
**Pull request by ClementKunz**
Syncing a folder containg lots of items inside it took a lot of time because of the status update of the parent folder

### CODE REVIEW

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 1]
```diff
@@ -1362,19 +1364,52 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
 
         if (isSyncing != isSyncingCurrent || isHydrated != isHydratedCurrent) {
             // Update parent directory status
-            QFileInfo info(path);
-            return vfsProcessDirStatus(info.dir().path());
+            const QString parentPath = QFileInfo(path).dir().path();
+            if (parentPath == _localSyncPath) return true;
+
+            if (isSyncing) {
+                {
```
Comment (luc-guyot-infomaniak) [inline src/server/vfs/mac/litesyncextconnector.mm:1371]: The indentation is odd.

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 2]
```diff
@@ -1405,20 +1440,23 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
             continue;
         }
 
-        bool isPlaceholder;
-        bool isHydrated;
-        bool isSyncing;
-        int progress;
-        if (!vfsGetStatus(tmpPath, isPlaceholder, isHydrated, isSyncing, progress)) {
+        bool isChildPlaceholder;
```
Comment (luc-guyot-infomaniak) [inline src/server/vfs/mac/litesyncextconnector.mm:1443]: Missing Initializations.

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 3]
```diff
@@ -1405,20 +1440,23 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
...
             continue;
         }
 
-        if (isSyncing) {
+        if (isChildSyncing) {
             existsOneStatusSync = true;
             break;
         }
 
-        existsOneStatusOnLine |= !isHydrated;
```
Comment (luc-guyot-infomaniak) [inline src/server/vfs/mac/litesyncextconnector.mm:1421]: You changed the logic here. Is it intended?

File: src/libsyncengine/syncpal/syncpal.cpp [hunk 4]
```diff
@@ -549,6 +549,14 @@ bool SyncPal::vfsForceStatus(const SyncPath &path, bool isSyncing, int progress,
     return _vfsForceStatus(_syncDbId, path, isSyncing, progress, isHydrated);
 }
 
+bool SyncPal::vfsCleanUpStatuses() {
+    if (!_vfsCleanUpStatuses) {
+        return false;
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/syncpal/syncpal.cpp:554]: Logging?

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 5]
```diff
@@ -1362,19 +1364,52 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
...
+                vfsSetStatus(parentPath, isSyncing, 100, isHydrated);
+            }
+            else {
+                _mutex.lock();
+                _syncingFolders[parentPath].remove(path);
+                if (_syncingFolders[parentPath].size() == 0) {
+                    _syncingFolders.remove(parentPath);
+                    _mutex.unlock();
+
+                    if (!vfsProcessDirStatus(parentPath)) {
```
Comment (luc-guyot-infomaniak) [inline src/server/vfs/mac/litesyncextconnector.mm:1384]: Should we adapt the other calls to `vfsProcessDirStatus` in this file?

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 6]
```diff
@@ -1362,19 +1364,52 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
...
         }
     }
     return true;
 }
 
+bool LiteSyncExtConnector::vfsCleanUpStatuses() {
+    const std::lock_guard<std::mutex> lock(_mutex);
+    QHashIterator<QString, QSet<QString>> it(_syncingFolders);
+    while (it.hasNext()) {
+        LOGW_WARN(_logger, L"TEST_CK - clean up needed for path=" << QStr2WStr(it.key()).c_str());
```
Comment (luc-guyot-infomaniak) [inline src/server/vfs/mac/litesyncextconnector.mm:1399]: To be removed?

File: src/server/vfs/mac/litesyncextconnector.mm (hunk 1 above)
Comment (ClementKunz) [reply src/server/vfs/mac/litesyncextconnector.mm:1371]: Yes but it is used to scope the mutex

File: src/server/vfs/mac/litesyncextconnector.mm [hunk 7]
```diff
@@ -1362,19 +1364,52 @@ bool checkIoErrorAndLogIfNeeded(IoError ioError, const std::string &itemType, co
...
             // Update parent directory status
-            QFileInfo info(path);
-            return vfsProcessDirStatus(info.dir().path());
+            const QString parentPath = QFileInfo(path).dir().path();
+            if (parentPath == _localSyncPath) return true;
+
+            if (isSyncing) {
+                {
+                    const std::lock_guard<std::mutex> lock(_mutex);
+                    _syncingFolders[parentPath].insert(path);
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply src/server/vfs/mac/litesyncextconnector.mm:1373]: Yes, agreed for later 👍

File: src/server/vfs/mac/litesyncextconnector.mm (hunk 3 above)
Comment (ClementKunz) [reply src/server/vfs/mac/litesyncextconnector.mm:1421]: YEs, if there is at least 1 dehydrated child, the parent must appear with the cloud icon.
The variables has been renamed though in order to improve readability.

File: src/libsyncengine/syncpal/syncpal.cpp (hunk 4 above)
Comment (ClementKunz) [reply src/libsyncengine/syncpal/syncpal.cpp:554]: I kept the same logic as the others callback (see all method starting with `vfs` in `syncpal.cpp`).

File: src/server/vfs/mac/litesyncextconnector.mm (hunk 5 above)
Comment (ClementKunz) [reply src/server/vfs/mac/litesyncextconnector.mm:1384]: Looks like the other call does not have any performance issue. I would prefer keep the changes to a minimum on this branch if the test does not show any problem with this implementation. What do you think?

File: src/server/vfs/mac/litesyncextconnector.mm (hunk 6 above)
Comment (ClementKunz) [reply src/server/vfs/mac/litesyncextconnector.mm:1399]: Oops, of course. Thx

**Review by luc-guyot-infomaniak — APPROVED** [review]

**Pull request merged**